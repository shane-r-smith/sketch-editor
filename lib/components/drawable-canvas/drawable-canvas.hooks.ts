import { useEffect, useMemo, useState } from "react";
import type { DrawableCanvasApi, Drawing } from "./drawable-canvas.types";
import type { DrawingState, SketchElement } from "../../domain";
import type { KonvaEventObject } from "konva/lib/Node";

const getSafeExitPosition = (
  position: number,
  min: number,
  max: number
): number => {
  if (position < min) {
    return min;
  }

  if (position > max) {
    return max;
  }

  return position;
};

export const useDrawing = ({
  tool,
  width,
  height,
  onDrawing,
  onDraw,
  disabled,
}: DrawableCanvasApi): Drawing => {
  const [drawingElement, setDrawingElement] = useState<SketchElement>();
  const [drawingState, setDrawingState] = useState<DrawingState>({
    isDrawing: false,
    hasLeftCanvasWhileDrawing: false,
  });

  const mouseUpOutsideListener = useMemo(() => {
    return () => {
      setDrawingElement(undefined);
      setDrawingState({ isDrawing: false, hasLeftCanvasWhileDrawing: false });
      window.removeEventListener("mouseup", mouseUpOutsideListener); // Clean-up
    };
  }, []);

  useEffect(
    /**
     * Effect to call the `onDrawing` property when `isDrawing` toggles.
     */
    () => {
      if (!onDrawing) {
        return;
      }

      onDrawing(drawingState);
    },
    [onDrawing, drawingState]
  );

  if (disabled) {
    return {};
  }

  const handleMouseDown = (e: KonvaEventObject<MouseEvent>) => {
    if (e.evt.button === 1) {
      return;
    }

    const pos = e?.target?.getStage()?.getPointerPosition();
    setDrawingState({
      isDrawing: true,
    });

    setDrawingElement({
      ...tool,
      points: pos ? [pos.x, pos.y] : [],
    });
  };

  const handleMouseMove = (e: KonvaEventObject<MouseEvent>) => {
    // No drawing - skipping
    if (!drawingState.isDrawing) {
      return;
    }

    if (!drawingElement) {
      return;
    }

    const stage = e.target.getStage();

    if (!stage) {
      return;
    }

    const point = stage.getPointerPosition();

    if (!point) {
      return;
    }

    setDrawingElement({
      ...drawingElement,
      points: drawingElement.points.concat([point.x, point.y]),
    });
  };

  const handleMouseUp = () => {
    if (!drawingElement) {
      setDrawingElement(undefined);
      setDrawingState({
        isDrawing: false,
      });
      return;
    }

    onDraw(drawingElement);

    setDrawingElement(undefined);
    setDrawingState({
      isDrawing: false,
    });
  };

  const handleMouseEnter = (e: KonvaEventObject<MouseEvent>) => {
    if (!drawingState.hasLeftCanvasWhileDrawing) {
      return;
    }

    // User reentered without lifting the mouse up. Remove event listener (reset to normal drawing state)
    window.removeEventListener("mouseup", mouseUpOutsideListener);

    // Renter drawing state
    setDrawingState({ isDrawing: true, hasLeftCanvasWhileDrawing: false });
    const pos = e?.target?.getStage()?.getPointerPosition();

    setDrawingElement({
      ...tool,
      points: pos ? [pos.x, pos.y] : [],
    });
  };

  const handleMouseLeave = (e: KonvaEventObject<MouseEvent>) => {
    if (!drawingState.isDrawing || !drawingElement) {
      return;
    }

    setDrawingState({ isDrawing: false, hasLeftCanvasWhileDrawing: true });

    // Save and stop drawing
    const pos = e?.target?.getStage()?.getPointerPosition();

    const x = pos ? getSafeExitPosition(pos.x, 0, width) : undefined;
    const y = pos ? getSafeExitPosition(pos.y, 0, height) : undefined;
    const xy = [x!, y!];

    onDraw({
      ...drawingElement,
      points: xy ? drawingElement.points.concat(xy) : drawingElement.points,
    });
    setDrawingElement(undefined);

    // Listen if the user lift's their mouse up
    window.addEventListener("mouseup", mouseUpOutsideListener);
  };

  return {
    drawingElement,
    onMouseDown: handleMouseDown,
    onMouseMove: handleMouseMove,
    onMouseUp: handleMouseUp,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onTouchStart: handleMouseDown as unknown as (
      evt: KonvaEventObject<TouchEvent>
    ) => void,
    onTouchMove: handleMouseMove as unknown as (
      evt: KonvaEventObject<TouchEvent>
    ) => void,
    onTouchEnd: handleMouseUp,
  };
};
