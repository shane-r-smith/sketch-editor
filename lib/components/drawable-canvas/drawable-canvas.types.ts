import type { Layer, SketchElement, Tool } from "../../domain";
import type { Stage } from "konva/lib/Stage";
import type { Ref } from "react";
import type { CanvasApi } from "../canvas";
import type { KonvaNodeEvents } from "react-konva";

export interface DrawingState {
  isDrawing: boolean;
  hasLeftWhileDrawing?: boolean;
}

export interface DrawableCanvasProps extends DrawableCanvasApi {
  ref: Ref<Stage>;
}

export interface DrawableCanvasApi extends CanvasApi {
  tool: Tool;
  currentLayerIndex: number;
  onDraw: (element: SketchElement) => void;
  onDrawing?: (state: DrawingState) => void;
  layers: Layer[];

  disabled?: boolean;
}

export interface Drawing
  extends Pick<
    KonvaNodeEvents,
    | "onMouseDown"
    | "onMouseMove"
    | "onMouseUp"
    | "onMouseEnter"
    | "onMouseLeave"
    | "onTouchStart"
    | "onTouchMove"
    | "onTouchEnd"
  > {
  drawingElement?: SketchElement;
}
