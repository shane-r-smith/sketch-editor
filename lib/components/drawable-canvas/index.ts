export * from "./drawable-canvas";
export * from "./drawable-canvas.hooks";
export * from "./drawable-canvas.types";
