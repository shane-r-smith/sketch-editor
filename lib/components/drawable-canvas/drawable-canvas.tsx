import type { DrawableCanvasProps } from "./drawable-canvas.types";
import { Canvas } from "../canvas";
import { useDrawing } from "./drawable-canvas.hooks";

export function DrawableCanvas(props: DrawableCanvasProps) {
  const drawing = useDrawing(props);

  if (props.disabled) {
    /** If disabled, render non-drawable version of the canvas */
    return <Canvas {...props} />;
  }

  return <Canvas {...props} {...drawing} />;
}
