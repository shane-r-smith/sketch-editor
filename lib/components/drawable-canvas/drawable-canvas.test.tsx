import { render } from "@testing-library/react";
import { DrawableCanvas } from "./drawable-canvas";
import type { DrawableCanvasProps } from "./drawable-canvas.types";
import { vi } from "vitest";

// Mock Canvas and useDrawing
vi.mock("../canvas", () => ({
  Canvas: (props: any) => (
    <div data-testid="canvas">{JSON.stringify(props, undefined, 2)} </div>
  ),
}));
vi.mock("./drawable-canvas.hooks", () => ({
  useDrawing: vi.fn(() => ({ drawingProp: "test" })),
}));

describe("<DrawableCanvas />", () => {
  const baseProps: DrawableCanvasProps = {
    width: 300,
    height: 150,
    disabled: false,
    ref: null,
    tool: {
      tool: "PEN",
      icon: "pen",
      stroke: "#FF0000",
      strokeWidth: 5,
      strokeColourType: "HEX",
      strokeColourHistory: [],
    },
    currentLayerIndex: 0,
    onDraw: () => {},
    layers: [],
  };

  it("matches snapshot when enabled", () => {
    const { container } = render(<DrawableCanvas {...baseProps} />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot when disabled", () => {
    const props = { ...baseProps, disabled: true };
    const { container } = render(<DrawableCanvas {...props} />);
    expect(container).toMatchSnapshot();
  });
});
