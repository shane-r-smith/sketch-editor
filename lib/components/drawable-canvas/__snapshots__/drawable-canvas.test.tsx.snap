// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DrawableCanvas /> > matches snapshot when disabled 1`] = `
<div>
  <div
    data-testid="canvas"
  >
    {
  "width": 300,
  "height": 150,
  "disabled": true,
  "ref": null,
  "tool": {
    "tool": "PEN",
    "icon": "pen",
    "stroke": "#FF0000",
    "strokeWidth": 5,
    "strokeColourType": "HEX",
    "strokeColourHistory": []
  },
  "currentLayerIndex": 0,
  "layers": []
}
     
  </div>
</div>
`;

exports[`<DrawableCanvas /> > matches snapshot when enabled 1`] = `
<div>
  <div
    data-testid="canvas"
  >
    {
  "width": 300,
  "height": 150,
  "disabled": false,
  "ref": null,
  "tool": {
    "tool": "PEN",
    "icon": "pen",
    "stroke": "#FF0000",
    "strokeWidth": 5,
    "strokeColourType": "HEX",
    "strokeColourHistory": []
  },
  "currentLayerIndex": 0,
  "layers": [],
  "drawingProp": "test"
}
     
  </div>
</div>
`;
